Dockerfile

```diff
@@ -1,22 +1,75 @@
-FROM python:3.12
+FROM ghcr.io/astral-sh/uv:python3.12-bookworm AS build-backend-stage
 
-ENV PYTHONDONTWRITEBYTECODE=1
-ENV PYTHONUNBUFFERED=1
+RUN export DEBIAN_FRONTEND=noninteractive && \
+    apt-get update && \
+    apt-get install -y \
+    -o APT::Install-Recommends=false \
+    -o APT::Install-Suggests=false \
+    libsasl2-dev \
+    libgraphviz-dev \
+    graphviz
 
-RUN apt-get update && apt-get install -y libsasl2-dev libgraphviz-dev graphviz
+WORKDIR /app
 
-COPY requirements.txt /kelvin/requirements.txt
-RUN pip install -r /kelvin/requirements.txt
+COPY pyproject.toml uv.lock ./
 
-RUN apt-get update \
-      && apt-get install -y ca-certificates curl gnupg lsb-release \
-      && curl -fsSL https://download.docker.com/linux/debian/gpg | gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg \
-      && echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/debian $(lsb_release -cs) stable" > /etc/apt/sources.list.d/docker.list \
-      && apt-get update \
-      && apt-get install -y docker-ce docker-ce-cli containerd.io
+RUN --mount=type=cache,target=/root/.cache uv sync --frozen --no-dev --no-install-project --compile-bytecode
 
-WORKDIR /kelvin
+FROM node:22.9.0-bookworm-slim AS build-frontend-stage
 
-#RUN addgroup --gid 1000 user
-#RUN adduser --disabled-password --gecos '' --uid $HOST_USER_ID --gid 1000 user
-#USER user
+WORKDIR /frontend
+
+COPY frontend .
+
+RUN mkdir -p /web/static
+
+RUN npm run build
+
+FROM python:3.12-bookworm AS runtime
+
+RUN export DEBIAN_FRONTEND=noninteractive && \
+    apt-get update && \
+    apt-get install -y \
+    -o APT::Install-Recommends=false \
+    -o APT::Install-Suggests=false \
+    graphviz && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
+WORKDIR /app
+
+# Create new user to run app process as unprivilaged user
+RUN groupadd --gid 1001 uvicorn && \
+    useradd --uid 1001 --gid 1001 --shell /bin/false --system uvicorn
+
+RUN chown -R uvicorn:uvicorn /app
+
+COPY --from=build-backend-stage --chown=uvicorn:uvicorn /app .
```

Comment:
Can't we do just `COPY . .` to avoid all this? E.g. something like this:
```dockerfile
COPY project.toml uv.lock .
RUN uv sync

COPY --from=frontend frontend/frontend.[js|css]|dolos

COPY --chown=uvicorn:uvicorn . .

...
```
In this way if frontend changes, we don't need to rebuild Python deps, but the Dockerfile should be much simpler. I don't think that we need to optimize the fast local Dockerfile rebuild case anyway though.

Reply:
I did this on purpose to find out which folders are not needed, so you can comment :D, but yes we can copy the whole folder.

Reply:
Ok, let's keep it, at least we'll make sure that there's no ballast being included in the image.